### DISCUSSION TITLE
Possibility to associate roles with user groups

### DISCUSSION THREAD

**Discussion by Ferks-FK** [spatie/laravel-permission#2469] [Q&A] [answered]
Hello, I would like to ask the more experienced if it is possible to associate a group of users with the roles.
I already managed to make it work the standard way, which is a user being associated with a role.
But I also have the functionality of user groups, where basically that group can be associated with a specific role, where everyone who is in that group, will have the role of the group associated with them.

**Comment by drbyte** [comment]
The comment posted here: https://github.com/spatie/laravel-permission/issues/310#issuecomment-1643586836
... discusses applying the HasRoles trait to another model.

You could then check `$user->hasRole(foo) || $userGroup->hasRole(foo)` in your application, and perhaps clone the RoleMiddleware and customize it to check both cases.

I'd be curious to see what you come up with for your application!

**Reply by Ferks-FK** [reply]
I didn't quite understand the logic of "cloning the roleMiddleware", can you provide a code example?

**Reply by Ferks-FK** [reply]
About the my application, you can see a demo here: [SourceAdmin](https://demo.sourceadmin.net/)

For you user group permissions thing, just log in, go to admin, and groups.
My idea is to add a select of available roles there, and the users who are associated with this group, will have the role associated with them as well.

**Reply by drbyte** [reply]
(Aside: I was suggesting making a copy of this package's RoleMiddleware into your own application, and customizing the line that checks $user->hasAnyRole to check hasAnyRole on both the user and whichever group you have assigned to the user. I'm not sure how you'll be referring to the user's groups in any middleware, so that's up to you and your application.)

Instead of talking about middleware (although it has relevance), I should have emphasized: to use this package to authorize users via the group they're assigned to, you'll need to check $group->hasRole or $group->hasPermissionTo.

**Comment by drbyte** [comment]
Have you explored the Teams feature of this package? Perhaps that's all you need?

**Reply by Ferks-FK** [reply]
If I understand correctly, the teams feature is basically a team of users.
Let's say: Team Admin, where you can have 5 users who are admins, and when we add a role by passing a team ID, all users related to that team will have access to the role, am I right about this?

**Reply by drbyte** [reply]
When a role is defined for a team, yes, all of that team's members will have what that role provides. Team members don't necessarily ever need to know who else is "part of" that team, it could merely be for organizing/grouping things.

@erikn69 do you have additional thoughts that might help here?

**Comment by Ferks-FK** [comment]
A small question that has nothing to do with the original post.
When we use `$user->can(...)` within policies, what is taken into account in the check?
I mean, does it check user permissions through their `roles`, or `permissions` directly?
I ask this, because in my case, the user can have a role, but he can also have a direct permission assigned to him (let me know if this is a bad thing or not).
What does the package take into account in this case?

**Reply by drbyte** [reply]
> the user can have a role, but he can also have a direct permission assigned to him (let me know if this is a bad thing or not).

That's not a bad thing. As you're doing, it's best to primarily assign roles, but then if needed can add bespoke additional permissions.

> When we use $user->can(...) within policies, what is taken into account in the check?
I mean, does it check user permissions through their roles, or permissions directly?

`$user->can()` calls `checkPermissionTo()`, which checks both role-based and directly-assigned permissions.

https://github.com/spatie/laravel-permission/blob/2dc2787dd8b0b6ccba2e40a7620a0a1653f39ec3/src/PermissionRegistrar.php#L125-L129

https://github.com/spatie/laravel-permission/blob/2dc2787dd8b0b6ccba2e40a7620a0a1653f39ec3/src/Traits/HasPermissions.php#L233

**Reply by Ferks-FK** [reply]
Thank you for clearing my doubt.

**Comment by Ferks-FK** [comment, marked as answer]
I was able to implement permissions by user groups using my own custom logic, this is because I could not understand very well how to work with the `Teams` feature of the package.
Any day I can edit this post showing how it was done.
Thanks to all.
